lnwallet/channel.go

```diff
@@ -5956,15 +5972,116 @@ func (lc *LightningChannel) availableBalance() (lnwire.MilliSatoshi, int64) {
-	commitFee := filteredView.feePerKw.FeeForWeight(commitWeight)
-	if lc.channelState.IsInitiator {
-		ourBalance -= lnwire.NewMSatFromSatoshis(commitFee)
+	// Do the same calculation from the remote commitment point of view.
+	ourRemoteCommitBalance, _ := lc.availableCommitmentBalance(
+		htlcView, true,
+	)
+
+	// Return which ever balance is lowest.
+	if ourRemoteCommitBalance < ourLocalCommitBalance {
+		return ourRemoteCommitBalance, commitWeight
+	}
+
+	return ourLocalCommitBalance, commitWeight
+}
+
+// availableCommitmentBalance attempts to calculate the balance we have
+// available for HTLCs on the local/remote commitment given the htlcView. To
+// account for sending HTLCs of different sizes, it will take into account
+// whether the HTLCs will be manifested on the commitment, increasing the
+// commitment fee we must pay as an initiator, eating into our balance.
+func (lc *LightningChannel) availableCommitmentBalance(view *htlcView,
+	remoteChain bool) (lnwire.MilliSatoshi, int64) {
+
+	// Compute the current balances for this commitment. This will take
+	// into account HTLCs to determine the commit weight, which the
+	// initiator must pay the fee for.
+	ourBalance, theirBalance, commitWeight, filteredView := lc.computeView(
+		view, remoteChain, false,
+	)
+	feePerKw := filteredView.feePerKw
+
+	// Set the dustlimit based on who's commitment we are looking at.
+	dustlimit := lnwire.NewMSatFromSatoshis(
+		lc.channelState.LocalChanCfg.DustLimit,
+	)
+	if remoteChain {
+		dustlimit = lnwire.NewMSatFromSatoshis(
+			lc.channelState.RemoteChanCfg.DustLimit,
+		)
+	}
+
+	// Given the commitment weight, find the commitment fee in case of no
+	// added HTLC output.
+	baseCommitFee := lnwire.NewMSatFromSatoshis(
+		feePerKw.FeeForWeight(commitWeight),
+	)
+
+	// Also calculate the commitment fee in the case where we would add
+	// another HTLC large enough to create a new output on the commitment.
+	htlcCommitFee := lnwire.NewMSatFromSatoshis(
+		feePerKw.FeeForWeight(commitWeight + input.HTLCWeight),
+	)
+
+	// For an extra HTLC fee to be paid on our commitment, the HTLC must be
+	// large enough to make a non-dust HTLC timeout transaction.
+	htlcFee := lnwire.NewMSatFromSatoshis(
+		htlcTimeoutFee(feePerKw),
+	)
+
+	// On the remote commitment, the HTLC success transaction will be used.
+	if remoteChain {
+		htlcFee = lnwire.NewMSatFromSatoshis(
+			htlcSuccessFee(feePerKw),
+		)
+	}
+
+	// The HTLC output will be manifested on the commitment if it
+	// is non-dust after paying the HTLC fee.
+	nonDustHtlcAmt := dustlimit + htlcFee
+
+	switch {
+
+	// If we are the channel initiator, we must to subtract the commitment
+	// fee from our available balance.
+	case lc.channelState.IsInitiator:
+		switch {
+		// If our balance is high enough to pay for a non-dust HTLC and
+		// the resulting commitment fee, subtract this commitment fee
+		// to reflect what we have left for such payments.
+		case ourBalance >= htlcCommitFee+nonDustHtlcAmt:
```

Comment:
If in this case and the default case below, we'd just return `0` as available balance, could we prevent users from sending out micropayments that get their channel into an unusable state?

Reply:
Goo idea, done for the initiator!

I still keep this calculation around for the non-initiator, since we letting them send dust HTLCs moves us away from the unusable state.